---
title: "Tuberculosis Bulk RNA-seq Analysis"
format:
  pdf:
    toc: true
    number-sections: true
    geometry: "margin=0.5in"
    fontsize: 11pt
execute:
  warning: false
knitr:
  opts_chunk:
    fig.align: "center"
---

## Overview
Singhania et al. profiled whole‑blood gene expression from humans with active tuberculosis, latent TB infection (LTBI), and uninfected controls across multiple cohorts, using microarrays and RNA‑seq. Their main goal was to identify an optimal reduced gene set to serve as a diagnostic biomarker for active TB that did not pick up other diseases.

The Leicester RNA‑seq cohort (GSE107994) used here includes longitudinal whole‑blood samples from a low‑incidence setting, and consists of active TB patients (n = 53) and recent close contacts (n = 108). The close contacts were categorized as either Interferon-Gamma Release Assay (IGRA) negative (n = 50) or IGRA positive (n = 49), indicating the absence or presence of *Mycobacterium tuberculosis*, respectively. A subset of the initial close contact subjects were identified as having active TB during the longitudinal assessment (n = 9) based on microbiological confirmation.

The authors used a modular analysis framework to group co‑expressed genes into immune “modules” (e.g. type I interferon–inducible, B‑cell, and T‑cell modules). They showed that active TB is characterized by overabundance of type I interferon‑inducible transcripts and underabundance of IFNG/TBX21 and B‑/T‑cell–associated genes, and that some individuals with LTBI display “outlier” active‑TB‑like signatures, suggesting phenotypic heterogeneity and possible higher progression risk.

For this project intended to familiarize myself with bulk RNA-seq data, I focus first on a cross‑sectional baseline comparison between active TB, LTBI, and controls to reproduce key aspects of the transcriptional signature (e.g., interferon‑related changes), and then extend to longitudinal dynamics and heterogeneity within the LTBI and LTBI‑progressor groups.

Singhania A, Verma R, Graham CM, Lee J et al. A modular transcriptional signature identifies phenotypic heterogeneity of human tuberculosis infection. Nat Commun 2018 Jun 19;9(1):2308. [PMID: 29921861](https://pubmed.ncbi.nlm.nih.gov/29921861/)

## The Data
(download data from GSE, establish needed local directories, format data objects)

```{r}
#| label: Setup

library(tidyverse) # dplyr, tidyr, tibble, readr, ggplot2
library(readxl)
library(Biobase)
library(GEOquery)

dir.create(file.path("data_raw"), showWarnings = FALSE) # establish directory to raw data downloads

data_dir <- file.path("data_raw", "GSE107994") # establish path to specific study directory
```

### Download data from GSE
(describe procedure for downloading data from GSE)

```{r}
#| label: Download-counts

if (!dir.exists(data_dir) || length(list.files(data_dir)) == 0) { # checks for supplemental files from GSE107994
  getGEOSuppFiles("GSE107994", baseDir = "data_raw") # accesses GEO for supplemental files (gene counts for all samples)
}

counts_file <- list.files( # grabs the specific file name for the raw gene counts data
  path = data_dir,
  pattern = "Raw_counts.*Leicester.*\\.xlsx$",
  full.names = TRUE
)

counts <- readxl::read_xlsx(counts_file) # function to import .xlsx file as data table
```

(get the phenotypic metadata)

```{r}
#| label: Download-phenotypes

gse <- getGEO("GSE107994", GSEMatrix = TRUE) # grabs the series matrix from GEO
phenotypes <- Biobase::pData(gse[[1]]) |> as_tibble() # formats GSE download as tabular metadata
```

{{< pagebreak >}}

### Inspection and storage

(snapshot of counts and phenotypes tables, discuss the cases)

```{r}
#| label: Inspect-data

print(counts)

print(phenotypes)
```


```{r}
#| label: Store-data

dir.create("data_processed", showWarnings = FALSE) # create data_processed directory folder if it does not exist

write_tsv(phenotypes, file.path("data_processed", "GSE107994_phenotypes.tsv")) # save phenotypes locally
write_tsv(counts, file.path("data_processed", "GSE107994_counts_raw.tsv")) # save counts locally
```

{{< pagebreak >}}

## Analysis 1 - Baseline Cross-Section
(Active TB vs inactive. Baseline samples. No outliers. Both sexes)
(look for enrichment of type I IFN‑inducible genes. How does baseline expression differ between ATB and contacts, and do we recover the type I IFN signature?)
(outliers determined by authors from transcriptional analyses)

```{r}
#| label: Analysis-1-setup

phenotypes_baseline <- phenotypes |> # subset phenotypes to the columns of interest
  select(
    geo_accession,
    patient_id = 'patient_id:ch1',
    group = 'group:ch1',
    tb_disease_type = 'tb_disease_type:ch1',
    smear_result = 'smear_result:ch1',
    outlier = 'outlier:ch1',
    gender = 'gender:ch1',
    ethnicity = 'ethnicity:ch1',
    birthplace = 'birth_place:ch1',
    age_baseline = 'age_at_baseline_visit:ch1',
    timepoint_months = 'timepoint_months:ch1',
    visit_date = 'visit_date:ch1',
    title
) |>
  mutate( # apply some factor levels to multiple variables of interest
    group = factor(group, levels = c("Control", "Active_TB", "LTBI", "LTBI_Progressor")),
    tb_disease_type = factor(tb_disease_type),
    smear_result = factor(smear_result, levels = c("Negative", "Positive")),
    gender = factor(gender)
  ) |> 
  dplyr::filter( 
    outlier == "No", # filter the data to remove predetermined outliers ...
    timepoint_months == "Baseline" # ... and focus on those who had active/inactive TB at baseline
  )

write_tsv(phenotypes_baseline, file = file.path("data_processed", "phenotypes_baseline.tsv"))

counts_mat <- counts |> # need to convert counts to a simple matrix
  select(-Gene_name, -Gene_biotype) |>
  column_to_rownames("Genes") |>
  as.matrix()

colnames(counts_mat) <- phenotypes$geo_accession[match(colnames(counts_mat), phenotypes$title)] # replace with shorter names

counts_baseline <- counts_mat[, phenotypes_baseline$geo_accession] # select only baseline samples
```

{{< pagebreak >}}

### Preprocessing
(describe intro to limma workflow)

```{r}
#| label: Analysis-1-preprocess

library(edgeR) # limma loaded as dependency

dge <- DGEList(counts = counts_baseline) # create DGEList object from raw counts matrix

keep_genes <- filterByExpr(dge, group = phenotypes_baseline$group) # Determine which genes have sufficiently large counts to be retained in a statistical analysis

dge <- dge[keep_genes, ] # retains only the genes determined to have sufficiently large counts

dge <- calcNormFactors( # Calculate scaling factors to convert the raw library sizes into normalized effective library sizes
  dge,
  method = "TMM" # trimmed mean of M-values
) # the data are not normalized yet. The factors are just stored in dge$samples$norm.factors for later use
```

(discuss DGE object and norm factors, not yet applied)

```{r}
#| label: Display-dge
#| echo: false

print(head(dge$samples))
```

(code commentary)

```{r}
#| label: Analysis-1-design

design_mat <- model.matrix(~ 0 + group, data = phenotypes_baseline) # prepare the design

colnames(design_mat) <- levels(phenotypes_baseline$group) # adjust names

voom_res <- voom(counts = dge, design = design_mat, plot = TRUE) # transform count data to log2 counts-per-million (logCPM)
  # incorporates the calc norm factors
```

(code commentary) 

{{< pagebreak >}}

### Principal Components Analysis

```{r}
#| label: Analysis-1-PCA

pca <- prcomp( # perform PCA
  x = t(voom_res$E), # must transpose data to samples x variables format
  center = TRUE,
  scale. = TRUE
)

pca$rotation |>
  as.data.frame() |>
  rownames_to_column("gene") |>
  write_csv(file = file.path("data_processed", "analysis1_pca_loadings.csv"))

pca_df <- pca$x |>
  as.data.frame() |>
  rownames_to_column("case")

write_csv(pca_df, file = file.path("data_processed", "analysis1_pca_scores.csv"))

prop_var <- pca$sdev^2 / sum(pca$sdev^2) * 100 # calculaute proportion of variance per PC

dir.create(file.path("output", "plots"), showWarnings = FALSE, recursive = TRUE)

library(cowplot)

pca_plot <- pca_df |>
  left_join(y = phenotypes_baseline, by = c("case" = "geo_accession")) |>
  ggplot(aes(x = PC1, y = PC2, color = group, shape = gender)) +
    geom_point(alpha = 0.75, size = 2) +
    theme_minimal_grid() +
    coord_fixed(ratio = 1) +
    scale_color_viridis_d(direction = -1, begin = 0.1, end = 1) +
    labs(
      title = "PCA of voom-normalized expression",
      x = paste0("PC1 (", round(prop_var[1], 2), "%)"),
      y = paste0("PC2 (", round(prop_var[2], 2), "%)"),
      subtitle = paste0(nrow(voom_res), " genes retained of ", nrow(counts), " genes measured")
    )

save_plot(
  filename = file.path("output", "plots", "analysis1_pca.png"),
  plot = pca_plot,
  bg = "white"
)

```

{{< pagebreak >}}

(describe PCA overview)
(noticeable contrast between active TB and the recent contact subjects)

```{r}
#| label: Analysis-1-PCA-plot
#| echo: false

pca_plot
```

{{< pagebreak >}}

### Fitting linear models
(uses original limma function)

```{r}
#| label: Analysis-1-limma

fit <- lmFit(
  object = voom_res,
  design = design_mat
)

print(head(coef(fit)))
```

(describe limma output)

```{r}
#| label: Analysis-1-contrasts

contrasts_mat <- makeContrasts(
  ATB_vs_allContacts = Active_TB - (Control + LTBI + LTBI_Progressor)/3,
  ATB_vs_Controls = Active_TB - Control,
  Progressors_vs_LTBI = LTBI_Progressor - LTBI,
  levels = design_mat
)

fit2 <- contrasts.fit(fit, contrasts_mat) |>
  eBayes()

print(head(coef(fit2)))
```

(describe additional steps of making contrasts, eBayes)

{{< pagebreak >}}

```{r}
#| label: Analysis-1-exports

dir.create(file.path("output", "tables"), showWarnings = FALSE, recursive = TRUE)

DE_genes <- list()

for (i in seq_along(colnames(contrasts_mat))) {

  top_table <- topTable(
    fit = fit2,
    coef = i,
    number = Inf,
    adjust.method = "BH",
    sort.by = "P"
  ) |>
    rownames_to_column("gene")
  
  top_table |>
    dplyr::filter(adj.P.Val < .05) |> # only export significant results
    write_csv(
      file = file.path(
        "output", "tables",
        paste0("analysis1_limma_", colnames(contrasts_mat)[i], ".csv")
      )
    )

  DE_genes[[colnames(contrasts_mat)[i]]] <- top_table

}

```

{{< pagebreak >}}

### Plotting differentially expressed features

```{r}
#| label: Analysis-1-plot-features

DE_genes_df <- bind_rows(DE_genes, .id = "contrast") # collapse list to data frame

DE_genes_plot <- DE_genes_df |>
  mutate(significant = adj.P.Val < .05 & abs(logFC) > 1) |>
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), color = significant)) +
    geom_point(size = 1, alpha = 0.8) +
    facet_wrap(~ contrast, nrow = 1) +
    theme_half_open() +
    theme(legend.position = "bottom") +
    scale_color_manual(values = c("FALSE" = "grey70", "TRUE" = "#48bda6")) +
    scale_y_continuous(expand = c(0,0), limits = c(0,NA)) +
    geom_hline(yintercept = -log10(0.05), linetype = "dotted", color = "grey40") +
    geom_vline(xintercept = 1, linetype = "dotted", color = "grey40") +
    geom_vline(xintercept = -1, linetype = "dotted", color = "grey40") +
    labs(
      x = "log2 fold change",
      y = "-log10 adj. p value",
      title = "Differentially expressed genes"
    )

cowplot::save_plot(
  filename = file.path("output", "plots", "analysis1_limma_contrasts.png"),
  plot = DE_genes_plot,
  bg = "white"
)

DE_genes_plot
```

## Analysis 2
(look at the longitudinal view within LTBI)
